### ISSUE TITLE
The behavior of assignment to LocalX variables doesn't match Windows

### ISSUE THREAD

**Issue by 8infy** [acpica/acpica#910]
The following code:
```
    Method (TEST)
    {
        Local0 = 0x64
        Local1 = RefOf(Local0)

        Local1 = 0x65
        
        Debug = Local0
        Debug = Local1
    }
```

Produces the following result on ACPICA:
```
ACPI Debug:  0x0000000000000064
ACPI Debug:  0x0000000000000065
```

Whereas on Windows 10 it produces the following trace:
```
Store(0x64,Local0)=0x64
Store(RefOf(Local0)=<Obj=FFFFA60201D90410,Type=DataAlias,Value=0xffffa60201d494d0,Buff=0000000000000000,Len=0>,Local1)=<Obj=FFFFA60201D49090,Type=DataAlias,Value=0xffffa60201d494d0,Buff=0000000000000000,Len=0>
Store(0x65,Local1)=0x65
Store(Local0=0x65,Debug)Integer(:Value=0x65[101])
=0x65
Store(Local1=<Obj=FFFFA60201D90410,Type=DataAlias,Value=0xffffa60201d494d0,Buff=0000000000000000,Len=0>,Debug)DataAlias(:Link=1d494d0)| Integer(:Value=0x65[101])
=<Obj=FFFFA60201D49090,Type=DataAlias,Value=0xffffa60201d494d0,Buff=0000000000000000,Len=0>
```

As you can see ACPICA overwrote the Local1 itself, whereas Acpi.sys overwrote the value referenced by Local1.

**Comment by 8infy** [comment]
> Hi Daniil,
> 
> Thanks for opening all the Issues highlighting behavior differences between ACPICA, Windows acpi.sys and the ACPI (6.5+) Specification. I will be back from vacation onwards of 12/18/2023 and will take a detailed look at all the suggestions, but looks like we might need to discuss all the behavior differences first internally and then open ECRs in ASWG to have the ACPI Specification language changed or removed/added to make sure all the 3 entities mentioned above are on the same page. I will make sure you get the credit for those ECRs if I have to open them on Mantis. Thanks for the good work and appreciate your patience!
> 
> -Saket

No problem, I'm glad these are useful. I've been looking more into the differences between ACPICA and the NT implementation while implementing my own interpreter, and there's just so many differences, some subtle, some not so much. What would be the best way to go about this? Is this something ACPICA would be interested in fixing or working out with the spec writers? I'm not familiar with what an ECR is or ASWG (nor do I know what Mantis is), but I'd like to help make ASL/AML more standardized across implementations.

**Comment by sacdintel** [comment]
Yes, we would like to fix implementation differences across ACPICA, Windows and the ACPI Specification or at the very least, address them to have converging behavior at some point in the future. You can keep on reporting all those differences here under "Issues" as you continue to discover more of them, but if you work for a company that is a UEFI Forum member then you can ask to be added to the ASWG (ACPI Specification Working Group) meetings (where you can bring these issues in discussions yourself) where the participants are representatives across the industry from various companies (chip design, OS vendors, Firmware/BIOS vendors, OEMs/ODMs etc.) including all the big names like Intel, Microsoft, Arm, Qualcomm, Nvidia, AMD, Google, Dell, Lenovo, HP, Apple etc. It is our goal to serve the open-source community and have ASL/AML more standardized across implementations and ACPICA tries to play nice by attempting to match closely other behaviors (closed-source Windows most notably). Discussions at ASWG meetings are usually kicked-off when developers open ECRs (Engineering Change Requests) on Mantis (online reporting and bug tracking utility akin to GitHub) within the UEFI Forum website. Thanks a lot once again for noting all the implementation differences and bringing it to our attention, we can take the next appropriate steps in the coming days to weeks and see the reaction at ASWG meetings.

**Comment by 8infy** [comment]
Thanks for the detailed response!

>if you work for a company that is a UEFI Forum member then you can ask to be added to the ASWG (ACPI Specification Working Group) meetings

I do this as my personal personal project on the side for fun so that's not a possibility unfortunately 😄

**Comment by sacdintel** [comment]
That's awesome! By the way, if you are willing to and also if you think a lot of new value would be unlocked, we can look into (after some discussions) adding your personal projects related to ACPI (like your interpreter which might address some of our ACPICA pain points or include new features) here, although I would first be required to look into how this could be turned into a possibility for both parties and sort out licensing headaches or ACPICA project policies. We can possibly carry your custom tools if they might add a lot of BIOS/Firmware developer productivity. No guarantees yet, just a possibility for now! It would be incredible for everyone to create more win-win situations!

**Comment by 8infy** [comment]
> That's awesome!

Thank you!

> By the way, if you are willing to and also if you think a lot of new value would be unlocked, we can look into (after some discussions) adding your personal projects related to ACPI (like your interpreter which might address some of our ACPICA pain points or include new features) here, although I would first be required to look into how this could be turned into a possibility for both parties and sort out licensing headaches or ACPICA project policies.

That would definitely be amazing!

>  We can possibly carry your custom tools if they might add a lot of BIOS/Firmware developer productivity.

Well, all that I have right now is just the interpreter that is half a year of relatively active development old, probably supporting about 70% of the language, although all of the core features are implemented. The biggest one missing ATM is `Field` support, the rest is misc instructions like ToString/Integer, Concat etc.
I also have a bunch of tests, but that's besides the point.

I try to research every possible edge case and look at the behavior or existing implementations before implementing every instruction, so it takes a while even for simpler stuff. Often times the specification ends up very detached from reality.

> No guarantees yet, just a possibility for now! It would be incredible for everyone to create more win-win situations!

Sure, sounds great to me. I was surprised at how drastic some of differences across implementations are. If we can improve that I'd be really happy.

**Comment by 8infy** [comment]
> That's awesome! By the way, if you are willing to and also if you think a lot of new value would be unlocked, we can look into (after some discussions) adding your personal projects related to ACPI (like your interpreter which might address some of our ACPICA pain points or include new features) here, although I would first be required to look into how this could be turned into a possibility for both parties and sort out licensing headaches or ACPICA project policies. We can possibly carry your custom tools if they might add a lot of BIOS/Firmware developer productivity. No guarantees yet, just a possibility for now! It would be incredible for everyone to create more win-win situations!

A small progress update from me: the last few months have been pretty active, and I'm happy to report that my interpreter now supports pretty much the entirety of AML! I already have a person that has integrated my interpreter into their kernel and was able to successfully enter sleep state 5 on four different devices, some PCs, some laptops (as well as VMs of course)!

There's still a lot to do of course, e.g. GED/_GPE are still missing. But it's coming together really well.

I can definitely say that your support and encouragement has given me more motivation to keep this up and for that I'm very grateful! ❤️

**Comment by sacdintel** [comment]
Thanks a lot for your kind words, it means a lot to me! I hope you are successful in all your endeavors including the repos you are working on! I did present all your issues (ACPICA vs acpi.sys) internally at Intel and in the next few weeks I hope to present it to the ASWG where all of them hopefully end up being approved! I will keep you and Armin posted on that! Thanks for your patience, I am back from a 6 week bonding leave for caring for my newborn!

**Comment by d-tatianin** [comment]
That's amazing, congratulations! Also very happy to hear my issues weren't just annoying spam, everyone eventually getting on the same page spec wise is great!

**Comment by sacdintel** [comment]
Thanks a lot!
